### PR TITLE
Elastic 2

### PR DESCRIPTION
**Pull request by pulse14**
I'm not sure what we should do for the ->query method. We can't predict what the user will pass and we need a bool query for simple filter.

And I'm not sure why there isn't options anymore on regex query on elastica. In the doc there is flags https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-regexp-query.html.

I've also temporarily addedan index refresh in delete tests. https://www.elastic.co/guide/en/elasticsearch/plugins/2.0/delete-by-query-plugin-reason.html . Seems to be needed now

### CODE REVIEW

**Comment by lorenzo** [comment]
@pulse14 I think we could also have a `search()` method. It will work exactly the same as `where` but instead of generating the filter conditions, it generates the query part for the `query()` method. What do you think?

**Comment by pulse14** [comment]
@lorenzo Where do we put the query part ?

**Comment by lorenzo** [comment]
No need to call `query()`. When compiling the query in `compileQuery()` we use the accumulated value of `search()` like we do with `where()`. The `query()` method can be left there to inject any `AbstractQuery` the user built themself

File: src/Query.php [hunk 1]
```diff
@@ -522,17 +520,17 @@ public function compileQuery()
...
-        $filteredQuery = new FilteredQuery();
-
-        if ($this->_parts['query'] !== null) {
-            $filteredQuery->setQuery($this->_parts['query']);
-            $this->_elasticQuery->setQuery($filteredQuery);
-        }
+//        if ($this->_parts['query'] !== null) {
+//            $filteredQuery->setQuery($this->_parts['query']);
+//            $this->_elasticQuery->setQuery($filteredQuery);
+//        }
```
Comment (markstory) [inline src/Query.php]: Commented out code should just be removed.

File: tests/TestCase/TypeTest.php [hunk 2]
```diff
@@ -568,6 +572,10 @@ public function testDeleteAll()
     public function testDeleteAllOnlySome()
     {
         $result = $this->type->deleteAll(['body' => 'cake']);
+        
+        //TODO : Not Sure
+        ConnectionManager::get('test')->getIndex()->refresh();
```
Comment (markstory) [inline tests/TestCase/TypeTest.php]: Forcing a refresh seems reasonable given that we can't rely on elastic search merging the new segments quickly.

**Comment by pulse14** [comment]
Ok I see. For example I add my query with `search()` and the query builder. And then I want to add a custom query with `query()` that **should** match, how do I do ? because If we stack all on the must clause this not gonna work.

```
{
  "query": {
    "bool": {
      "must": [
        {
          "bool": {
            "should": {
              "term": {
                "test": "test"
              }
            }
          }
        },
        {
          "bool": {
            "should": {
              "term": {
                "value": "b"
              }
            }
          }
        }
      ]
    }
  }
}
```

is not equivalent to

```
{
  "query": {
    "bool": {
      "must": [
        {
          "bool": {
            "should": {
              "term": {
                "test": "test"
              }
            }
          }
        }
      ],
      "should": {
        "bool": {
          "should": {
            "term": {
              "value": "b"
            }
          }
        }
      }
    }
  }
}
```
I don't know if the user expect such behavior. I think we should not stack all in the must cause

**Comment by lorenzo** [comment]
I my mind, using `query` overrides whatever was passed in search. Either that or we remove the `query()` method entirely

**Comment by pulse14** [comment]
@lorenzo Maybe query could simply override the query completly except postFilter. It gives more flexibility

**Comment by lorenzo** [comment]
@pulse14 sounds good to me

**Comment by pulse14** [comment]
Not sure why elastica removed options for regex

```php
/**
     * Construct regexp query.
     *
     * @param string $key   OPTIONAL Regexp key
     * @param string $value OPTIONAL Regexp value
     * @param float  $boost OPTIONAL Boost value (default = 1)
     */
    public function __construct($key = '', $value = null, $boost = 1.0)
    {
        if (!empty($key)) {
            $this->setValue($key, $value, $boost);
        }
    }
```
We can do one manually but I think there is a reason

**Comment by lorenzo** [comment]
Let's adjust to what elastica has @pulse14. We can remove the options as well

**Comment by lorenzo** [comment]
I'm going to merge this as it is a great start, and then try to improve on it by sending more PRs.

Thanks a lot @pulse14

**Pull request merged**